### ISSUE TITLE
issue with the kernel expected when running Elyra pipeline

### ISSUE THREAD

**Issue by pacospace** [elyra-ai/elyra#1373]
**Describe the issue**
I created a new kernel (e.g. `python3new` ) and added it to the notebook kernel, therefore it is part of the notebook metadata and it can be found from command:
```
[1000600000@jupyterhub-nb-fmurdaca-40redhat-2ecom python3]$ jupyter kernelspec list
Available kernels:
  python3-test    /opt/app-root/src/.local/share/jupyter/kernels/python3-test
  python3new      /opt/app-root/src/.local/share/jupyter/kernels/python3new
  python3         /opt/app-root/share/jupyter/kernels/python3
```

Then when submitting an Elyra pipeline the kernel selected is not found.

Looking at the code my understanding is that on the image used as base runtime for the step in the pipeline that kernel name actually does not exists, so if the kernel name from notebook metadata is selected it will fail as it cannot find kernel.json file (https://github.com/elyra-ai/kfp-notebook/blob/a1f1a9d4ba21b1d2b664afce440095f1f0912cd1/etc/docker-scripts/bootstrapper.py#L413). Switching back to `python3` kernel name in notebook metadata let the step run without problems on the runtime image.

Is it an expected behaviour? I would expect that the new-kernel name (the dependencies are provided by the runtime image used for the step so there is no problem) which is part of the notebook metadata does not cause failure in the step. Moreover, the issue identified is a bit ambigous because the kernel.json for the actual kernel name is the file missing.

**To Reproduce**
1. Create a new kernel `python3new` with `jupyterlab-requirements` [[2]] for example
2. check the kernel is available `jupyter kernelspec list`
3. Assign the kernel to the notebook
4. Run an Elyra pipeline with runtime image that is compatible.

**Screenshots or log output**
download_dataset.ipynb
[I 17:02:58.610] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - dependencies processed (1.161 secs)
[I 17:02:58.611] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - executing notebook using 'papermill download_dataset.ipynb download_dataset-output.ipynb' 
[D 17:02:59.327] Native kernel (python3) available from /opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources
[E 17:02:59.327] Unexpected error: <class 'FileNotFoundError'>
[I 17:02:59.457] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - converting from download_dataset-output.ipynb to download_dataset.html 
Traceback (most recent call last):
  File "bootstrapper.py", line 341, in execute
    kernel_name = NotebookFileOp.find_best_kernel(notebook)
  File "bootstrapper.py", line 415, in find_best_kernel
    with open(os.path.join(file, 'kernel.json')) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources/kernel.json'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/app-root/src/jupyter-work-dir/python3/nbformat/__init__.py", line 138, in read
    buf = fp.read()
AttributeError: 'str' object has no attribute 'read'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "bootstrapper.py", line 594, in <module>
    main()
  File "bootstrapper.py", line 584, in main
    file_op.execute()
  File "bootstrapper.py", line 356, in execute
    NotebookFileOp.convert_notebook_to_html(notebook_output, notebook_html)
  File "bootstrapper.py", line 374, in convert_notebook_to_html
    nb = nbformat.read(notebook_file, as_version=4)
  File "/opt/app-root/src/jupyter-work-dir/python3/nbformat/__init__.py", line 140, in read
    with io.open(fp, encoding='utf-8') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'download_dataset-output.ipynb'
```

**Expected behavior**
No failures in the Elyra pipeline.

**Deployment information**
Describe what you've deployed and how: [[1]]
 - Elyra version: 2.* [[1]]
 - Operating system: linux [[1]]
 - Installation source: [[1]]
 - Deployment type: Open Data Hub on Openshift, Elyra image spawn from JH and Kubeflow + Tekton as engine for pipelines run. [[1]]

[1]: https://github.com/thoth-station/elyra-aidevsecops-tutorial/blob/master/README.md#environment-required
[2]: https://github.com/thoth-station/jupyterlab-requirements

**Comment by kevin-bates** [comment]
Unless the new kernel configuration (i.e., its kernelspec directory) is available, and locatable, on the image used in KFP, the bootstrapper will go through its search of a language-matching kernel - which is evident in the traceback.  However, that kernel match encountered an issue and I suspect its due to the image's configuration.

Is this a custom image being used?
Is the image available for inspection?  If not, can you please run `jupyter kernelspec list` _from within that particular image_ and provide the results?  The kernel-match logic essentially uses the `KernelSpecManager` that `jupyter kernelspec list` uses to identify a match.

**Comment by pacospace** [comment]
> Unless the new kernel configuration (i.e., its kernelspec directory) is available, and locatable, on the image used in KFP, the bootstrapper will go through its search of a language-matching kernel - which is evident in the traceback. However, that kernel match encountered an issue and I suspect its due to the image's configuration.
> 
> Is this a custom image being used?
> Is the image available for inspection? If not, can you please run `jupyter kernelspec list` _from within that particular image_ and provide the results? The kernel-match logic essentially uses the `KernelSpecManager` that `jupyter kernelspec list` uses to identify a match.

Thanks @kevin-bates for the detailed explanation. 

The image used is available here: quay.io/thoth-station/elyra-aidevsecops-tutorial:v0.5.0 but does not has any jupyter package (it was built with software stack specific for the notebook). isn't `jupyter` installed from bootstraper.py in kfp-notebook?

**Comment by kevin-bates** [comment]
>  isn't jupyter installed from bootstraper.py in kfp-notebook library

Good point - although we don't install a kernel instance - we only [pip install the ipykernel package, among others](https://github.com/elyra-ai/kfp-notebook/blob/master/etc/requirements-elyra.txt). That said, we clearly need to firm up the potential for a missing kernel.json in the kernelspec directory, but the fact that a kernelspec directory was found w/o a kernel.json file violates what I had figured was an invariant condition.  I'll take a look at the image - thanks for the information.

**Comment by kevin-bates** [comment]
**TL;DR:** We're not handling "native kernel" fallback logic properly and your image introduced a scenario that both my test with your image or any other scenarios had yet to uncover.  That said, if your pipeline is dependent on your custom kernel configuration, you will need to ensure that that kernelspec is resident in your image and discoverable via the best-match logic.

Hmm - this is a bit odd.  Here's what I've done...
First, I attempted to run `jupyter kernelspec list` directly against your image and got the expected result...
```
$ docker run -it --rm quay.io/thoth-station/elyra-aidevsecops-tutorial:v0.5.0 jupyter kernelspec list
/usr/bin/container-entrypoint: line 2: exec: jupyter: not found
```
Next, I extended the image with the previously linked `elyra-requirements.txt` file using this Dockerfile:
```
$ cat Dockerfile 
FROM quay.io/thoth-station/elyra-aidevsecops-tutorial:v0.5.0

COPY requirements-elyra.txt /tmp
RUN python -m pip install -r /tmp/requirements-elyra.txt
```
and this `docker build` command...
```
$ docker build . -t elyra-aidevsecops-w-reqs
```
It turns out a plain installation of `ipykernel` does install a kernespec...
```
$ docker run -it --rm elyra-aidevsecops-w-reqs jupyter kernelspec list
Available kernels:
  python3    /opt/app-root/share/jupyter/kernels/python3
```
and the kernelspec directory contains a properly formatted `kernel.json` file...
```
~/dev/docker/elyra-custom$ docker run -it --rm elyra-aidevsecops-w-reqs cat /opt/app-root/share/jupyter/kernels/python3/kernel.json
{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```
I also get the same directory using code that is used in the best-kernel-match logic (just wanted to make sure the CLI command didn't produce anything unusual):
```python
from jupyter_client.kernelspec import KernelSpecManager

def best_kernel():
    kernel_specs = KernelSpecManager().find_kernel_specs()
    for name, file in kernel_specs.items():
        print(f"{name}: {file}")

if __name__ == "__main__":
    best_kernel()
```
However, another clue is this DEBUG statement:
```
[D 17:02:59.327] Native kernel (python3) available from /opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources
```
This DEBUG is printed from the [`KernelSpecManager().find_kernel_specs()`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/kernelspec.py#L171-L178) method when the [`NATIVE_KERNEL_NAME`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/kernelspec.py#L24) is not already in the results list.  Since that value is `python3`, this implies that, for whatever reason, the installation of the elyra-requirements did NOT create a valid kernelspec.  And, since the 'file' element of the result set points at the [`RESOURCES`](https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelspec.py#L21) directory of where `ipykernel` is installed, the matching logic doesn't handle that (and it should).  In essence, whenever the "native kernel" is used as a fallback, the `kernel.json` is essentially created on the fly.  As a result, when the bootstrapper encounters this condition (which I would express as a) `kernel.json` is missing and b) the basename of the directory is `resources`) then we should treat this as a python-language kernel for matching purposes.  If that's a language match, we're done - else move on.

What I don't understand is why the creation of the proper kernelspec worked in my scenario, but not yours.  I suspect it might have something to do with where python is located in the path etc.  Since the Native kernel is found in bowels of `jupyter-work-dir` under `/opt/app-root/src/jupyter-work-dir`, I suspect the jupyter path logic used by Jupyter code is not finding the non-native (and expected) kernelspec definition and only the fallback - or the non-native spec if failing to be installed during the requirements processing.  

In fact, I see that my experiment probably isn't valid since we adjust the [working python lib directories](https://github.com/elyra-ai/kfp-notebook/blob/master/kfp_notebook/pipeline/_notebook_op.py#L134-L140) via the Notebook operator.

At any rate, since the change for this issue needs to come from the kfp-notebook repo, I'm going to transfer this there.

cc: @akchinSTC
